internal/plugin_test.go

```diff
@@ -1145,3 +1145,215 @@ subjects:
 	expected = strings.TrimPrefix(expected, "\n")
 	assertEqual(t, p.outputBuffer.String(), expected)
 }
+
+func TestGeneratePolicySets(t *testing.T) {
+	t.Parallel()
+	tmpDir := t.TempDir()
+	createConfigMap(t, tmpDir, "configmap.yaml")
+	p := Plugin{}
+	var err error
+	p.baseDirectory, err = filepath.EvalSymlinks(tmpDir)
+	if err != nil {
+		t.Fatal(err.Error())
+	}
+	p.PlacementBindingDefaults.Name = "my-placement-binding"
+	p.PolicyDefaults.Placement.Name = "my-placement-rule"
+	p.PolicyDefaults.Namespace = "my-policies"
+	patch := map[string]interface{}{
+		"metadata": map[string]interface{}{
+			"labels": map[string]string{
+				"chandler": "bing",
+			},
+		},
+	}
+	policyConf := types.PolicyConfig{
+		Name: "policy-app-config",
+		Manifests: []types.Manifest{
+			{
+				Path:    path.Join(tmpDir, "configmap.yaml"),
+				Patches: []map[string]interface{}{patch},
+			},
+		},
+	}
+	policyConf2 := types.PolicyConfig{
+		Name: "policy-app-config2",
+		Manifests: []types.Manifest{
+			{Path: path.Join(tmpDir, "configmap.yaml")},
+		},
+	}
+	p.Policies = append(p.Policies, policyConf, policyConf2)
+
+	// PolicyDefaults.PolicySets should be applied to both policies
+	p.PolicyDefaults.PolicySets = []string{"policyset-default"}
+	p.applyDefaults(map[string]interface{}{})
+	assertReflectEqual(t, p.Policies[0].PolicySets, []string{"policyset-default"})
+	assertReflectEqual(t, p.Policies[1].PolicySets, []string{"policyset-default"})
+	expectedPolicySets := []types.PolicySetConfig{
+		{
+			Name: "policyset-default",
+			Policies: []string{
+				"policy-app-config",
+				"policy-app-config2",
+			},
+		},
+	}
+	assertReflectEqual(t, p.PolicySets, expectedPolicySets)
+	p.PolicySets = nil
+	// p.PolicyDefaults.PolicySets should be overridden by p.Policies[0].PolicySets
+	p.Policies[0] = types.PolicyConfig{
+		Name: "policy-app-config",
+		Manifests: []types.Manifest{
+			{
+				Path:    path.Join(tmpDir, "configmap.yaml"),
+				Patches: []map[string]interface{}{patch},
+			},
+		},
+		PolicySets: []string{"policyset0"},
+	}
+	p.applyDefaults(map[string]interface{}{})
```

Comment:
For every call you make to `p.applyDefaults`, could you please make it a separate test with a for range loop and `t.Run`? I'd like to have individual pass/fail results for each of these tests. It lets me know which testcase failed when viewing the test results rather than looking at the line that failed.

Reply:
sure